File: rust/src/latex.rs
//! LaTeX formula extraction module (R1).
//!
//! Extracts LaTeX formulas from HTML content, handling multiple sources:
//! - Habr: `img.formula` elements with `source` attribute
//! - `KaTeX`: `.katex` elements with `annotation[encoding="application/x-tex"]`
//! - `MathJax`: `mjx-container` elements with `data-tex`/`data-latex` attributes
//!
//! Based on reference implementation from:
//! <https://github.com/link-foundation/meta-theory/blob/main/scripts/download-article.mjs>

use scraper::{ElementRef, Selector};

/// Check if an element is a formula image (Habr-specific).
///
/// Habr renders formulas as SVG/PNG images with class `formula`.
#[must_use]
pub fn is_formula_image(element: &ElementRef) -> bool {
    let value = element.value();
    if value.name() != "img" {
        return false;
    }
    let classes = value.attr("class").unwrap_or("");
    classes.contains("formula") || value.attr("source").is_some()
}

/// Check if an element is a math element (`KaTeX`, `MathJax`, or generic math class).
#[must_use]
pub fn is_math_element(element: &ElementRef) -> bool {
    let value = element.value();
    let tag = value.name();
    let classes = value.attr("class").unwrap_or("");
    classes.contains("katex")
        || classes.contains("math")
        || classes.contains("MathJax")
        || tag == "mjx-container"
}

/// Extract LaTeX source from a formula image element (Habr-specific).
///
/// Habr renders formulas as SVG/PNG images with class `formula`.
/// The original LaTeX source is in the `source` attribute.
#[must_use]
pub fn extract_habr_formula(element: &ElementRef) -> Option<String> {
    let value = element.value();
    if let Some(source) = value.attr("source") {
        let trimmed = source.trim();
        if !trimmed.is_empty() {
            return Some(trimmed.to_string());
        }
    }
    if let Some(alt) = value.attr("alt") {
        let trimmed = alt.trim();
        if !trimmed.is_empty() {
            return Some(trimmed.to_string());
        }
    }
    None
}

/// Extract LaTeX from `KaTeX` elements.
///
/// `KaTeX` stores the TeX source in `annotation[encoding="application/x-tex"]`.
#[must_use]
pub fn extract_katex_formula(element: &ElementRef) -> Option<String> {
    // Look for annotation element
    if let Ok(sel) = Selector::parse(r#"annotation[encoding="application/x-tex"]"#) {
        if let Some(annotation) = element.select(&sel).next() {
            let text: String = annotation.text().collect();
            let trimmed = text.trim();
            if !trimmed.is_empty() {
                return Some(trimmed.to_string());
            }
        }
    }
    // Fallback to data-tex or data-latex attributes
    let value = element.value();
    if let Some(tex) = value.attr("data-tex").or_else(|| value.attr("data-latex")) {
        let trimmed = tex.trim();
        if !trimmed.is_empty() {
            return Some(trimmed.to_string());
        }
    }
    None
}

/// Extract LaTeX from `MathJax` elements.
///
/// `MathJax` stores TeX in `data-tex` attribute or annotation elements.
#[must_use]
pub fn extract_mathjax_formula(element: &ElementRef) -> Option<String> {
    let value = element.value();
    // First try data-tex/data-latex attributes
    if let Some(tex) = value.attr("data-tex").or_else(|| value.attr("data-latex")) {
        let trimmed = tex.trim();
        if !trimmed.is_empty() {
            return Some(trimmed.to_string());
        }
    }
    // Fallback to annotation element
    if let Ok(sel) = Selector::parse(r#"annotation[encoding="application/x-tex"]"#) {
        if let Some(annotation) = element.select(&sel).next() {
            let text: String = annotation.text().collect();
            let trimmed = text.trim();
            if !trimmed.is_empty() {
                return Some(trimmed.to_string());
            }
        }
    }
    None
}

/// Extract formula from any supported element type.
#[must_use]
pub fn extract_formula(element: &ElementRef) -> Option<String> {
    if is_formula_image(element) {
        return extract_habr_formula(element);
    }
    let tag = element.value().name();
    if tag == "mjx-container" {
        return extract_mathjax_formula(element);
    }
    if is_math_element(element) {
        return extract_katex_formula(element);
    }
    None
}

File: js/src/latex.js
/**
 * LaTeX formula extraction module.
 *
 * Extracts LaTeX formulas from HTML content, handling multiple sources:
 * - Habr: img.formula elements with `source` attribute
 * - KaTeX: .katex elements with annotation[encoding="application/x-tex"]
 * - MathJax: mjx-container elements with data-tex/data-latex attributes
 *
 * Based on reference implementation from:
 * https://github.com/link-foundation/meta-theory/blob/main/scripts/download-article.mjs
 */

/**
 * Extract LaTeX source from a formula image element (Habr-specific).
 * Habr renders formulas as SVG/PNG images with class "formula".
 * The original LaTeX source is in the `source` attribute.
 *
 * @param {Object} $ - Cheerio instance
 * @param {Object} el - Cheerio element (img.formula)
 * @returns {string|null} LaTeX source or null
 */
export function extractHabrFormula($, el) {
  const source = $(el).attr('source');
  if (source) {
    return source.trim();
  }
  const alt = $(el).attr('alt');
  if (alt) {
    return alt.trim();
  }
  return null;
}

/**
 * Extract LaTeX from KaTeX elements.
 * KaTeX stores the TeX source in annotation[encoding="application/x-tex"].
 *
 * @param {Object} $ - Cheerio instance
 * @param {Object} el - Cheerio element (.katex or .math)
 * @returns {string|null} LaTeX source or null
 */
export function extractKatexFormula($, el) {
  const annotation = $(el).find('annotation[encoding="application/x-tex"]');
  if (annotation.length > 0) {
    return annotation.text().trim();
  }
  const dataTex = $(el).attr('data-tex') || $(el).attr('data-latex');
  if (dataTex) {
    return dataTex.trim();
  }
  return null;
}

/**
 * Extract LaTeX from MathJax elements.
 * MathJax stores TeX in data-tex attribute or annotation elements.
 *
 * @param {Object} $ - Cheerio instance
 * @param {Object} el - Cheerio element (mjx-container)
 * @returns {string|null} LaTeX source or null
 */
export function extractMathJaxFormula($, el) {
  const dataTex = $(el).attr('data-tex') || $(el).attr('data-latex');
  if (dataTex) {
    return dataTex.trim();
  }
  const annotation = $(el).find('annotation[encoding="application/x-tex"]');
  if (annotation.length > 0) {
    return annotation.text().trim();
  }
  return null;
}

/**
 * Check if an element is a formula image (Habr-specific).
 *
 * @param {Object} $ - Cheerio instance
 * @param {Object} el - Cheerio element
 * @returns {boolean}
 */
export function isFormulaImage($, el) {
  return (
    $(el).is('img') &&
    ($(el).hasClass('formula') ||
      $(el).attr('source') !== undefined ||
      ($(el).attr('class') || '').includes('formula'))
  );
}

/**
 * Check if an element is a math element (KaTeX, MathJax, or generic math class).
 *
 * @param {Object} $ - Cheerio instance
 * @param {Object} el - Cheerio element
 * @returns {boolean}
 */
export function isMathElement($, el) {
  const tag = (el.tagName || el.name || '').toLowerCase();
  const classes = $(el).attr('class') || '';
  return (
    classes.includes('katex') ||
    classes.includes('math') ||
    classes.includes('MathJax') ||
    tag === 'mjx-container'
  );
}

/**
 * Extract formula from any supported element type.
 *
 * @param {Object} $ - Cheerio instance
 * @param {Object} el - Cheerio element
 * @returns {string|null} LaTeX source or null
 */
export function extractFormula($, el) {
  if (isFormulaImage($, el)) {
    return extractHabrFormula($, el);
  }
  const tag = (el.tagName || el.name || '').toLowerCase();
  if (tag === 'mjx-container') {
    return extractMathJaxFormula($, el);
  }
  if (isMathElement($, el)) {
    return extractKatexFormula($, el);
  }
  return null;
}
